go/vt/vtctl/reparentutil/durability.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reparentutil
+
+import (
+	"fmt"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/vtctl/reparentutil/promotionrule"
+)
+
+//=======================================================================
+
+// A newDurabler is a function that creates a new durabler based on the
+// properties specified in the input map. Every durabler must
+// register a newDurabler function.
+type newDurabler func(map[string]string) durabler
+
+var (
+	// durabilityPolicies is a map that stores the functions needed to create a new durabler
+	durabilityPolicies = make(map[string]newDurabler)
+	// curDurabilityPolicy is the current durability policy in use
+	curDurabilityPolicy durabler
+)
+
+func init() {
+	// register all the durability rules with their functions to create them
+	registerDurability("none", func(map[string]string) durabler {
+		return &durabilityNone{}
+	})
+	registerDurability("semi_sync", func(map[string]string) durabler {
+		return &durabilitySemiSync{}
+	})
+	registerDurability("cross_cell", func(map[string]string) durabler {
+		return &durabilityCrossCell{}
+	})
+	registerDurability("specified", newDurabilitySpecified)
+}
+
+// durabler is the interface which is used to get the promotion rules for candidates and the semi sync setup
+type durabler interface {
+	promotionRule(*topodatapb.Tablet) promotionrule.CandidatePromotionRule
+	primarySemiSync(*topodatapb.Tablet) int
+	replicaSemiSync(primary, replica *topodatapb.Tablet) bool
```

Comment:
ah, i see in the top-level functions later, but this is still confusing (because they are named the same i would have expected them to return the same thing, but just one for primaries and one for replicas)

Reply:
PrimarySemiSync is the number of replicas that need to send an ACK before the write is accepted.
ReplicaSemiSync is a boolean which is whether the replica can send a semi-sync or not. We can rename it if you have a better idea

Reply:
I agree that the names are confusing. We can rename `primarySemiSync` to be more descriptive of its functionality.